Configuration Manager

hqiu-at-wiris opened the pull request:
- Add utils subpackage.
- Implement configuration manager  and tests.

hqiu-at-wiris committed:
KB-4851: Add utils subpackage. Implement configuration manager  and tests.

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+from typing import List, Tuple, Dict, Any
+
+
+class UndefinedSetting(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+
+
+class UndefinedConfiguration(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+
+
+class InvalidConfiguration(Exception):
+    def __init__(self, message):
+        super().__init__(message)
```
I'm wondering if it is a good idea to set the messages that these exceptions pass in stone i.e. hardcoding them or parameterizing them. For example, for `UndefinedSetting` we really just need to pass the setting that was attempted to access/set and use `f"Setting {setting} has not been defined"` as the error message.

I say this because I think in this instance it would reduce code duplication and make same-exception messages uniform, but I'm not aware if this is a good practice in general. What do you think @dmas-at-wiris @HENRYQIULO ?

dmas-at-wiris replied:
I agree with @didac-fernandez

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+
+            for configuration in predefined_configurations:
+
+                if "NAME" not in configuration and "CONFIG" not in configuration:
```
Should be `or`, right?

dmas-at-wiris replied:
Yes, De Morgan's laws.

dmas-at-wiris replied:
@HENRYQIULO @didac-fernandez Random question. Why did we put NAME and CONFIG in uppercase?

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+            else:
+                # Set the first configuratoin as default.
+                self.apply_configuration(name=next(iter(self._configurations)))
+        elif default_configuration:
```
why is this an `else`? Does it not make sense to have both predefined configurations and defaults at the same time? If instead we simply try to set the given default using 
```python
if default_configuration:
    self.apply_configuration(name=default_configuration)
```
if the configuration has not been defined, the method will automatically raise the error for us. This would work well, unless we want to have a specific treatment for the default configuration

dmas-at-wiris replied:
@HENRYQIULO can you explain this case?

hqiu-at-wiris replied:
As both parameters were optional , this was just to prevent a user giving a default configuration without giving a predefined configuration, I think we can just ignore this weird case.

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+        Initialize internal structures that store the configurations.
+
+        A list of `predefined_configurations` can be passed to initialze the configuration manager,
+        together with the name of the `default_configuration`, if it's not given, he first
```
Due to the difficulty of tracking dictionary key order (which wasn't even guaranteed very few Python versions ago), I would make it such that if this `default_configuration` is not specified then no configuration is applied and that is left for the user to do. The class should throw exceptions if the user tries to use it without an active configuration.

Additionally, the usage of the word `default` in this context confuses me. I'm seeing two possible uses for the name:
- The default configuration is the one we enable ONLY the first time the configuration_manager is created. This would mean that, by construction, we have the property that there is always an active configuration, which is desirable.
- The default configuration is a special configuration. One that gets enabled by default and one where if something goes wrong we can fall back to the default (here this means we would need to store that configuration's name in a variable and have methods like `set_default_configuration`, `apply_default_configuration` or `get_default_configuration`). Currently I do not see a use for this.

At the very least, I would name the first one (the one being used here) along the lines of `initial_configuration`, not to be confused with the second.

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+        except KeyError:
+            raise UndefinedConfiguration(f"Configuration '{name}' undefined.")
+
+    def get_current_configuration(self) -> str:
```
Suggestion to rename anything 'current' to 'active'

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+
+        self._settings: Dict[str, Any] = {}
+        self._configurations: Dict[str, Dict[str, Any]] = {}
+        self._current_configuration: str = ""
```
Those are class variables and should be outside the init, without being initialized, except maybe by setting `_current_configuration` to `None` so that we can make simple checks and throw exceptions if there is no active configuration. I would also suggest renaming this to `active_configuration`.

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+
+    def list_settings(self) -> List[Tuple[str, str]]:
+        """
+        Return all pairs of key-value settings in the current configuration.
```
Avoid saying return in the docstring, it's redundant. Maybe something like 'Generate a list of all key-value pairs...'. Also, since settings can be changed independently, I would say 'a list of the currently active settings' rather than the ones 'in the current configuration'.

We should be careful with settings being editable in general. It seems pretty pythonic to me to leave it as is as long as everyone is aware this can happen, without enforcing more stringent conditions (like e.g. whenever a setting is edited the variable `_current_configuration` is nulled as it no longer represents the current state).

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+
+    def set_setting(self, key: str, value: Any) -> None:
+        """
+        Add a pair key-value or replace the value of a setting in the current configuration.
```
typo: "pair key-value" -> "key-value pair"

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+
+        self._settings[key] = value
+
+    def define_configuration(self, name: str, configuration: Dict[str, Any]) -> None:
```
Add an `override: bool = False` parameter. If override is set to False, throw exception if the configuration already exists.

dfernandez-at-wiris commented on santoku/utils/test_configuration_manager.py:
```diff
@@ -0,0 +1,152 @@
+
+
+@pytest.fixture(scope="class")
+def boolean_setting():
```
Should probably be a key-value pair? That's the definition we gave to `setting`. Named tuples seem like a good fit here:
```python
>>> from collections import namedtuple
>>> Setting = namedtuple("Setting", "name value")
>>> boolean_setting = Setting("boolean_setting", True)
>>> boolean_setting.name
'boolean_setting'
>>> boolean_setting.value
True
>>> type(boolean_setting)
<class '__main__.Setting'>
```
Similarly with configurations, I can see how it would make the code more parsable to do `configuration.settings` than `configuration["CONFIG"]`

dfernandez-at-wiris commented on santoku/utils/test_configuration_manager.py:
```diff
@@ -0,0 +1,152 @@
+    def test_set_setting(self, boolean_setting, default_configuration, configuration_manager):
+        # Change the value of an already defined setting in the current configuration.
+        # Success expected.
+        expected_value = default_configuration["CONFIG"][boolean_setting]
```
I do not understand this test, there is no call to the function being tested, is there?

Also, to keep tests independent, maybe we should avoid using other functions in the test other than the one being tested. In this case, instead of using `get_setting`, we could go a bit lower level and access the internal state of the configuration_manager class itself (i.e. the `settings` dictionary)

Same thing goes for most of the test suite.

dfernandez-at-wiris commented on santoku/utils/test_configuration_manager.py:
```diff
@@ -0,0 +1,152 @@
+        # Add a configuration to an empty configuration manager and get its settings.
+        # Success expected.
+        configuration_manager = ConfigurationManager()
+        new_configuration = {
```
this looks like a fixture to me

dfernandez-at-wiris commented on santoku/utils/configuration_manager.py:
```diff
@@ -0,0 +1,264 @@
+
+
+class ConfigurationManager:
+    """
+    Class to handle feature toggles (flags) in a project.
+
+    Feature flags can be used to set values for some required parameters in the project or to enable
+    or disable different parts of the source code that will be run in different executions. We
+    define the concept of `setting` as each individual flag that controls certain feature, i.e.
+    enabling/disabling the mock of an external service. A setting is a pair of key-value, being
+    `key` the name of the setting, and `value` an object of any type. We also define the concept of
+    `configuration` as a collection of multiple settings. In a project, each configuration should be
+    formed by settings containing the same keys but different values. The purpose of this class is
+    to handle the definition and retrieving of all the configuartion parameters so that the user can
+    control the execution context of the code.
+
+    """
```
Suggestion:
```python
"""
The purpose of this abstract class is to centrally manage configuration parameters for an entire application.

We define one of those parameters as a `setting`. A `setting` is a key-value pair, where they key is a string (its unique ID) and the value can be an object of any type. The following are all valid settings:

"boolean_setting": True
"numeric_setting": 42
"text_setting": "A long time ago in a galaxy far, far away..."
"object_setting": {
	"key1": True,
	"key2": 4,
	...
}

A `configuration` is a named collection of settings, taking the form of a dictionary with the following structure:

{
	"NAME": <configuration name (str, unique)>,
	"SETTINGS": {
		"setting1": <value1>,
		"setting2": <value2>,
		...	
	}
}
"""
```

hqiu-at-wiris committed:
- Add namedtuple to fixtures.
- Update docstring.
- Add editconfiguration method.

dfernandez-at-wiris committed:
#taskid 4851: WIP

hqiu-at-wiris committed:
#taskid 4851: Complete ConfigurationManager
- get rid of complex recursive structures
- add support for JSON Schema
- complete docstrings

hqiu-at-wiris committed:
#taskid 4851: Tests completed.

dfernandez-at-wiris committed:
#taskid 4851: Refactor, better fixtures, bugfix
- better fixture names
- more atomic usage of some fixtures
- fix dictionary copies. dict.copy() produces a shallow copy and thus the original object is modified in some circumstances, copy.deepcopy() performs the desired operation

dfernandez-at-wiris committed:
#taskid 4851: Encapsulate tests (to follow convention)

dfernandez-at-wiris committed:
Cosmetic update

dmas-at-wiris commented on santoku/utils/README.md:
````diff
@@ -0,0 +1,111 @@
+JSON = Union[List["JSON"], Dict[str, Union[bool, int, float, str, "JSON"]]
+```
+
+Since this issue is currently [the most demanded from the mypy project](https://github.com/python/mypy/issues/731), it is expected that we will eventually have support for recursive type checking. However, even if mypy raises and error, the code above does compile. For now, we will have to make do with the less powerful (i.e. wrong) version:
````
Why is highlighting this paragraph as if it was code?

dfernandez-at-wiris replied:
Does not happen in VSCode so it is a GitHub thing. It might be that the three dots in the JSON-style code block get it confused?? Idk...

dmas-at-wiris commented on santoku/utils/configuration.py:
```diff
@@ -0,0 +1,347 @@
+
+        Parameters
+        ----------
+        configurations : List[Dict[str, Union[str, Dict[str, Any]]]], optional
```
🤯

dmas-at-wiris commented on santoku/utils/configuration.py:
```diff
@@ -0,0 +1,347 @@
+        https://json-schema.org/specification-links.html#draft-7
+        """
+        if schema:
+            Draft7Validator.check_schema(schema=schema)
```
I guess this is throwing an exception. We leave that one to bubble?

dfernandez-at-wiris replied:
Yes, all validators do nothing if validation is successful and throw exceptions if the validation fails. All in all it makes sense to let their more useful error messages bubble than to capture them.

dmas-at-wiris commented on santoku/utils/configuration.py:
```diff
@@ -0,0 +1,347 @@
+            raise NoActiveConfiguration
+        return self.get_configuration(name=self.active_configuration)
+
+    def get_setting(self, key: Any) -> Any:
```
Let's imagine this case:
{
  "a": 1.
  "b": {
    "c": 2,
    "d": 3
  }
}

Can I query b and thus retrieving the dictionary containing c and d?

hqiu-at-wiris replied:
Sure! It's already supported.

dfernandez-at-wiris replied:
Well, we thought of ways to be able to "lazily" serve values that are not in a leaf node, but all of them were (very) terrible. We were convinced that trying to retrieve many values deep in the nesting if we didn't allow you to get intermediate values would be very tedious, so we ended up letting the user get intermediate results.

Now... I just realized that you can do something like:

```JSON
{
    "a": {
        "b": {
            "c": {
                "d": {
                    "e": 1,
                    "f": 1,
                    "g": 1
                }
            }
        }
    }
}
```

```python
keys_to_get_to_penultimate_level = ["a", "b", "c", "d"]
leaf_node_keys = ["e", "f", "g"]
deeply_nested_settings = {}
for key in leaf_node_keys:
    deeply_nested_settings[key] = config_manager.get_setting(keys_to_get_to_penultimate_level + [key])
```

so it is not too bad after all for a worse case scenario. Thus, I now think it makes sense to except if the user wants to access intermediate levels of the hierarchy and make them put in the work to extract many deeply nested values at once. This logic can even be very easily abstracted into a function.

dfernandez-at-wiris committed:
#taskid 4851: Forbid non-leaf node settings access
- add a new exception for this particular case
- sort exceptions alphabetically

dfernandez-at-wiris committed:
#taskid 4851: Fix & test ilegal access exception
- fix bug in the throwing of the exception
- implement unit test for that exception
